### ISSUE TITLE
Unarchive module throws an error on OS X

### ISSUE THREAD

**Issue by anthonygreen** [ansible/ansible-modules-core#2423]
#### Issue Type:

Bug Report
#### Ansible Version:

2.0.0-0.4.beta2
#### Environment

Running from: OS X 10.9.5
Managing: OSX VM 10.10
#### Summary:

unarchive_module throws an error on OS X
#### Steps to reproduce

This task fails

```
- name: Download a file
  unarchive: 
    src: https://www.dropbox.com/path.to.some/file?raw=1
    dest: /tmp
    copy: no
    creates: /tmp/file
```

results in the error

```
No such file or directory: '/tmp/__MACOSX/._
```

The file has been downloaded to /tmp and appears to unzipped but has thrown an error.

**Comment by robynbergeron** [comment]
[module: files/unarchive.py]

**Comment by robynbergeron** [comment]
@anthonygreen Thanks for submitting this bug report.

@pileofrogs thanks in advance for taking a look at this bug report and for your ongoing work in maintaining this module.

**Comment by pileofrogs** [comment]
I don't have acces to an OSX box.  I'll see if I can find anything obvious to fix.  What was the name and extension and filetype of the file in question?

**Comment by bcoca** [comment]
@anthonygreen also if you could run with -vvvv and show the output, it would help with the debugging

**Comment by anthonygreen** [comment]
The issue persists in ansible 2.0.0-0.6.rc1

<localhost> ESTABLISH SSH CONNECTION FOR USER: vagrant
<localhost> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o Port=2200 -o 'IdentityFile=".vagrant/machines/jupiter/virtualbox/private_key"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 -o ControlPath=/Users/greena28/.ansible/cp/ansible-ssh-%h-%p-%r -tt localhost (umask 22 && mkdir -p "`echo $HOME/.ansible/tmp/ansible-tmp-1448819403.73-45890821819007`" && echo "`echo $HOME/.ansible/tmp/ansible-tmp-1448819403.73-45890821819007`")
<localhost> ESTABLISH SSH CONNECTION FOR USER: vagrant
<localhost> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o Port=2200 -o 'IdentityFile=".vagrant/machines/jupiter/virtualbox/private_key"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 -o ControlPath=/Users/greena28/.ansible/cp/ansible-ssh-%h-%p-%r -tt localhost (umask 22 && mkdir -p "`echo $HOME/.ansible/tmp/ansible-tmp-1448819403.78-121559500485647`" && echo "`echo $HOME/.ansible/tmp/ansible-tmp-1448819403.78-121559500485647`")
<localhost> PUT /var/folders/sm/vgbwtrps3yv2yl79n43t_f916ldxgl/T/tmpx7vdpP TO /Users/vagrant/.ansible/tmp/ansible-tmp-1448819403.78-121559500485647/stat
<localhost> SSH: EXEC sftp -b - -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o Port=2200 -o 'IdentityFile=".vagrant/machines/jupiter/virtualbox/private_key"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 -o ControlPath=/Users/greena28/.ansible/cp/ansible-ssh-%h-%p-%r [localhost]
<localhost> ESTABLISH SSH CONNECTION FOR USER: vagrant
<localhost> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o Port=2200 -o 'IdentityFile=".vagrant/machines/jupiter/virtualbox/private_key"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 -o ControlPath=/Users/greena28/.ansible/cp/ansible-ssh-%h-%p-%r -tt localhost LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /Users/vagrant/.ansible/tmp/ansible-tmp-1448819403.78-121559500485647/stat; rm -rf "/Users/vagrant/.ansible/tmp/ansible-tmp-1448819403.78-121559500485647/" > /dev/null 2>&1
<localhost> ESTABLISH SSH CONNECTION FOR USER: vagrant
<localhost> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o Port=2200 -o 'IdentityFile=".vagrant/machines/jupiter/virtualbox/private_key"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 -o ControlPath=/Users/greena28/.ansible/cp/ansible-ssh-%h-%p-%r -tt localhost rc=flag; [ -r /tmp/ ] || rc=2; [ -f /tmp/ ] || rc=1; [ -d /tmp/ ] && rc=3; python -V 2>/dev/null || rc=4; [ x"$rc" != "xflag" ] && echo "${rc}  "/tmp/ && exit 0; (python -c 'import hashlib; BLOCKSIZE = 65536; hasher = hashlib.sha1();
afile = open("'/tmp/'", "rb")
buf = afile.read(BLOCKSIZE)
while len(buf) > 0:
    hasher.update(buf)
    buf = afile.read(BLOCKSIZE)
afile.close()
print(hasher.hexdigest())' 2>/dev/null) || (python -c 'import sha; BLOCKSIZE = 65536; hasher = sha.sha();
afile = open("'/tmp/'", "rb")
buf = afile.read(BLOCKSIZE)
while len(buf) > 0:
    hasher.update(buf)
    buf = afile.read(BLOCKSIZE)
afile.close()
print(hasher.hexdigest())' 2>/dev/null) || (echo '0  '/tmp/)
<localhost> ESTABLISH SSH CONNECTION FOR USER: vagrant
<localhost> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o Port=2200 -o 'IdentityFile=".vagrant/machines/jupiter/virtualbox/private_key"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 -o ControlPath=/Users/greena28/.ansible/cp/ansible-ssh-%h-%p-%r -tt localhost (umask 22 && mkdir -p "`echo $HOME/.ansible/tmp/ansible-tmp-1448819404.06-61884863236063`" && echo "`echo $HOME/.ansible/tmp/ansible-tmp-1448819404.06-61884863236063`")
<localhost> PUT /var/folders/sm/vgbwtrps3yv2yl79n43t_f916ldxgl/T/tmptstAHz TO /Users/vagrant/.ansible/tmp/ansible-tmp-1448819404.06-61884863236063/unarchive
<localhost> SSH: EXEC sftp -b - -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o Port=2200 -o 'IdentityFile=".vagrant/machines/jupiter/virtualbox/private_key"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 -o ControlPath=/Users/greena28/.ansible/cp/ansible-ssh-%h-%p-%r [localhost]
<localhost> ESTABLISH SSH CONNECTION FOR USER: vagrant
<localhost> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o Port=2200 -o 'IdentityFile=".vagrant/machines/jupiter/virtualbox/private_key"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=vagrant -o ConnectTimeout=10 -o ControlPath=/Users/greena28/.ansible/cp/ansible-ssh-%h-%p-%r -tt localhost LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /Users/vagrant/.ansible/tmp/ansible-tmp-1448819404.06-61884863236063/unarchive; rm -rf "/Users/vagrant/.ansible/tmp/ansible-tmp-1448819404.06-61884863236063/" > /dev/null 2>&1
fatal: [mars]: FAILED! => {"changed": false, "failed": true, "msg": "Unexpected error when accessing exploded file: [Errno 2] No such file or directory: '/tmp/__MACOSX/._Command_Line_Tools_OS_X_10.10_for_Xcode_7.1.dmg'", "stat": {"exists": false}}

**Comment by pileofrogs** [comment]
What was the archive?  I see you're getting it off dropbox, but I don't see what kind of archive it is.  EG is it tar, tgz, zip?

**Comment by anthonygreen** [comment]
@pileofrogs zip

**Comment by mcatst** [comment]
@anthonygreen I experienced this same problem using a tar file and found it to be my issue.  I wonder if your zip file issue is similar? When you create a tar file on Mac OS tar will package up the extended attributes of a file and store them in the tar file as another file with that '._' prefix. You can see them if you list the contents of the tar.  if you have a file 'textfile.txt' you'll also have a '._textfile.txt' in your archive.  The source for ansible unarchive, around line 334, starts iterating through the files in the archive and tries to manipulate them in the filesystem.  The problem is that those '.__' files don't exist in the same way which is causing the exception.  Rebuilding my tar using --exclude=".__" eliminated those from the tar and then everything was happy.  As to if this is a defect in ansible depends on if ansible desires to preserve the extended attributes.  I've seen no value in doing so yet.  I hope that helps.

**Comment by bcoca** [comment]
OS X fork files (the `._textfile.txt`) are a well known annoyance when dealing with shared drives or archives, but they should not cause errors as other systems will treat them as 'real files'.

**Comment by eirnym** [comment]
I have same issue with any archive on OS X/FreeBSD which contains or not `._filename` files.

The bug is unarchive.py when handler tries to check if file is already unpacked. `tar` command doesn't accept `-d` and `--diff` parameters so is_unarchive() function is useless :-) 

More so, execution result hasn't checked at all.

Please, add compatibility layer for FreeBSD/OSX.

My ansible version is `2.1.0.0`, but this bug is in master branch here.

**Comment by ansibot** [comment]
@dagwieers, @pileofrogs, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by dagwieers** [comment]
@all Please test this again with the latest **unarchive** module, and verify that you are using GNU tar and not BSD tar. Since the **unarchive** module does not support non-gnu tar. See #4352 for the details.

needs_info

**Comment by ansibot** [comment]
@anthonygreen we are unable to validate that "" is an existing ansible module.

We are expecting a component name that matches something that can be found in:
- http://docs.ansible.com/ansible/list_of_all_modules.html

Please revise the component name in the description.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dagwieers** [comment]
@anthonygreen Without any additional information, or a confirmation this is still a problem after refactoring the zip-support in unarchive, I am closing this issue as wontfix. Please re-open if this would still be a problem to you.

**Comment by eirnym** [comment]
@dagwieers Why not to add support for BSD tar? the only `--diff` may be problem

**Comment by dagwieers** [comment]
@eirnym We accept contributions. I don't think we would refuse a proper implementation that is backward compatible. But the current logic heavily relies on --diff and we don't want to have a generic poorer implementation just to support BSD tar (especially when GNU tar is available).

Again, a pure-python implementation is probably time better spent.

**Comment by eirnym** [comment]
@dagwieers I'll write it on pure Python. it's easier than support holy wars between BSD, POSIX and GNU versions.

**Comment by dagwieers** [comment]
It has nothing whatsoever to do with holy wars. I am afraid you are overlooking the complexity. It is not specific to the lack of `--diff`, we use some other options specific to GNU tar and we are scraping the output as well.

So the only pragmatic way to support this with one implementation is GNU tar, which is available on all UNIX platforms...

**Comment by eirnym** [comment]
I see your point. But I want to tell some disadvantages of this way:
- You can enforce dependency on provisioning machine, but you can't enforce dependency on target.
- DevOps will miss filesystem-specific flags like what exists in FreeBSD and OS X.
- Practically you use only two GNU-specific options: `--diff` and `--show-transformed-names` which are implementable
- I don't know why do you really need exact same output here. Almost all produced by Ansible Python scripts will throw this output away, other will know the output difference we had

What can be done to make all implementations possible (Pure-Python, BSD and GNU ways):
- With better interface and class organisation we can implement both with same functionality and output. We can wrap tar command itself, separate options and switches then call implementation-specific call.
- We definitely need to refactor this module (too long and too complex) including method separation. I don't think you cover 60% with tests.
- We need to write tests to be sure we are on same page.

I understand, that way I suggest will take some more time than say "we don't support". But implementation will be more strait forward than rejection.

**Comment by dagwieers** [comment]
@eirnym We are not just using `--diff` and `--show-transformed-names`, we also scrape the output and match it with regular expressions. I doubt that BSD tar would emit the exact same output. So yes, basically what you need to do is implement everything yourself. Mind you that the old implementation that did support both GNU tar _and_ BSD tar basically _always_ extracted everything. If you want that behaviour, sure that _is_ possible with little effort. Good luck because it is the same as doing `command: tar`. You don't need a module for that.

But again, if someone wants to do the required work I am not stopping you. Feel free to contribute whatever you think is needed. _I welcome contributions_, however the current implementation is GNU tar specific and that is what I stated, that is what the documentation states and it is the only thing we have today.

Want to change this for tomorrow's implementation, feel free to do that work. Open Source does not mean that someone else will write something for you, if you really want it, write it yourself !

**Comment by dagwieers** [comment]
Closing this ticket because
1. The issue is related to zip, while the discussion is about tar
2. We need more feedback wrt. the zip issue
3. The zip support has been redesigned (idempotent/check-mode/diff-mode) and likely fixes the issue in the devel branch.
bug_resolved

**Comment by eirnym** [comment]
@dagwieers according to code in master, you use these these options.

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.
